ballista/rust/core/src/utils.rs

```diff
@@ -54,9 +58,47 @@ use std::time::Duration;
 use std::{fs::File, pin::Pin};
 use tonic::codegen::StdError;
 use tonic::transport::{Channel, Error, Server};
+use url::Url;
 
-/// Stream data to disk in Arrow IPC format
+/// Default session builder using the provided configuration
+pub fn default_session_builder(config: SessionConfig) -> SessionState {
+    SessionState::with_config_rt(
+        config,
+        Arc::new(
+            RuntimeEnv::new(with_object_store_provider(RuntimeConfig::default()))
+                .unwrap(),
+        ),
+    )
+}
 
+/// Get a RuntimeConfig with specific ObjectStoreDetector in the ObjectStoreRegistry
+pub fn with_object_store_provider(config: RuntimeConfig) -> RuntimeConfig {
+    config.with_object_store_registry(Arc::new(ObjectStoreRegistry::new_with_provider(
+        Some(Arc::new(FeatureBasedObjectStoreProvider)),
+    )))
+}
+
+/// An object store detector based on which features are enable for different kinds of object stores
+pub struct FeatureBasedObjectStoreProvider;
+
+impl ObjectStoreProvider for FeatureBasedObjectStoreProvider {
+    /// Detector a suitable object store based on its url if possible
+    /// Return the key and object store
+    #[allow(unused_variables)]
+    fn get_by_url(&self, url: &Url) -> Option<Arc<dyn ObjectStore>> {
+        #[cfg(feature = "hdfs")]
```

Comment:
Being able to enable/disable object stores based on features sounds sensible, but I would recommend supporting multiple based on the scheme. Having to recompile different binaries for different backends is unfortunate. It's possible you intend to do this anyway, but thought I would mention

Reply:
For a standalone system, the binaries for the backend should be fixed before running. How would you use the capability of one crate without compiling them.

Reply:
For a standalone system, we should know which object stores to be used before running it. Therefore, it's OK to make the decision at the compiling phase.

Reply:
How would you distribute binaries for such a system?

Reply:
We run it in K8S with replicas.

Reply:
Let me phrase it differently, how would the ballista project distribute binaries/docker images for general consumption in arbitrary environments. I presume we want to provide this?

Reply:
We made an agreement that it's supposed to be able to find all of the object stores from the *ObjectStoreProvider*. Otherwise, an error should be thrown. We'll change the related ObjectStoreProvider interface after we upgrading to latest datafusion.